txmanager/pgstorage.go

```diff
@@ -0,0 +1,278 @@
+}
+
+// NewPostgresStorage creates a new instance of storage that use
+// postgres to store data
+func NewPostgresStorage(dbCfg db.Config) (*PostgresStorage, error) {
+	db, err := db.NewSQLDB(dbCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &PostgresStorage{
+		db,
+	}, nil
+}
+
+// Add persist a monitored tx
+func (s *PostgresStorage) Add(ctx context.Context, mTx txmTypes.MonitoredTx, dbTx pgx.Tx) error {
+	conn := s.dbConn(dbTx)
+	cmd := `
+        INSERT INTO state.monitored_txs (owner, id, from_addr, to_addr, nonce, value, data, gas, gas_offset, gas_price, status, block_num, history, created_at, updated_at, num_retries)
+                                 VALUES (   $1, $2,        $3,      $4,    $5,    $6,   $7,  $8,         $9,       $10,    $11,       $12,     $13,        $14,        $15,         $16)`
+
+	_, err := conn.Exec(ctx, cmd, mTx.Owner,
+		mTx.ID, mTx.From.String(), mTx.ToStringPtr(),
+		mTx.Nonce, mTx.ValueU64Ptr(), mTx.DataStringPtr(),
+		mTx.Gas, mTx.GasOffset, mTx.GasPrice.Uint64(), string(mTx.Status), mTx.BlockNumberU64Ptr(),
+		mTx.HistoryStringSlice(), time.Now().UTC().Round(time.Microsecond),
+		time.Now().UTC().Round(time.Microsecond), mTx.NumRetries)
+
+	if err != nil {
+		if pgErr, ok := err.(*pgconn.PgError); ok && pgErr.ConstraintName == "monitored_txs_pkey" {
+			return txmTypes.ErrAlreadyExists
+		} else {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Get loads a persisted monitored tx
+func (s *PostgresStorage) Get(ctx context.Context, owner, id string, dbTx pgx.Tx) (txmTypes.MonitoredTx, error) {
+	conn := s.dbConn(dbTx)
+	cmd := `
+        SELECT owner, id, from_addr, to_addr, nonce, value, data, gas, gas_offset, gas_price, status, block_num, history, created_at, updated_at, num_retries
+          FROM state.monitored_txs
+         WHERE owner = $1 
+           AND id = $2`
+
+	mTx := txmTypes.MonitoredTx{}
+
+	row := conn.QueryRow(ctx, cmd, owner, id)
+	err := s.scanMtx(row, &mTx)
+	if errors.Is(err, pgx.ErrNoRows) {
+		return mTx, txmTypes.ErrNotFound
+	} else if err != nil {
+		return mTx, err
+	}
+
+	return mTx, nil
+}
+
+// GetByStatus loads all monitored tx that match the provided status
+func (s *PostgresStorage) GetByStatus(ctx context.Context, owner *string, statuses []txmTypes.MonitoredTxStatus, dbTx pgx.Tx) ([]txmTypes.MonitoredTx, error) {
+	hasStatusToFilter := len(statuses) > 0
+
+	conn := s.dbConn(dbTx)
+	cmd := `
+        SELECT owner, id, from_addr, to_addr, nonce, value, data, gas, gas_offset, gas_price, status, block_num, history, created_at, updated_at, num_retries
+          FROM state.monitored_txs
+         WHERE (owner = $1 OR $1 IS NULL)`
+	if hasStatusToFilter {
+		cmd += `
+           AND status = ANY($2)`
+	}
+	cmd += `
+         ORDER BY created_at`
+
+	mTxs := []txmTypes.MonitoredTx{}
+
+	var rows pgx.Rows
+	var err error
+	if hasStatusToFilter {
+		rows, err = conn.Query(ctx, cmd, owner, statuses)
+	} else {
+		rows, err = conn.Query(ctx, cmd, owner)
+	}
+
```

Comment:
I realize you copied this code from elsewhere, but there should be a `defer rows.Close()` here

Reply:
Indeed. Good catch. Roman actually added this through another PR on the zkevm node. I'll add it here as well.